packages/next/src/client/components/app-router.tsx

```diff
@@ -416,13 +463,66 @@ function Router({
     use(createInfinitePromise())
   }
 
-  /**
-   * Handle popstate event, this is used to handle back/forward in the browser.
-   * By default dispatches ACTION_RESTORE, however if the history entry was not pushed/replaced by app-router it will reload the page.
-   * That case can happen when the old router injected the history entry.
-   */
-  const onPopState = useCallback(
-    ({ state }: PopStateEvent) => {
+  useEffect(() => {
+    if (process.env.__NEXT_WINDOW_HISTORY_SUPPORT) {
+      // Ensure the canonical URL in the Next.js Router is updated when the URL is changed so that `usePathname` and `useSearchParams` hold the pushed values.
+      const applyUrlFromHistoryPushReplace = (
+        url: string | URL | null | undefined
+      ) => {
+        startTransition(() => {
+          dispatch({
+            type: ACTION_RESTORE,
+            url: new URL(url ?? window.location.href),
```

Comment:
I created issue https://github.com/vercel/next.js/issues/56636 some time ago. I spotted this experimental feature in the 14.0.3 release notes and thought it could fix it. However, when I try it, I just get `TypeError: Failed to construct 'URL': Invalid URL` errors. It looks to be related to this review comment.

Reply:
@cduff see https://github.com/vercel/next.js/pull/58438

Reply:
Thanks @franky47. I just tested against `14.0.4-canary.1` which includes https://github.com/vercel/next.js/pull/58438 and it looks like it will now resolve https://github.com/vercel/next.js/issues/56636.